lib/parsec/validate.py

```diff
@@ -23,365 +23,511 @@
+                            raise IllegalItemError(
+                                keys, key, 'consecutive spaces')
+                        if ((val_is_dict and spc_is_dict) or
+                                (not val_is_dict and not spc_is_dict)):
+                            speckey = '__MANY__'
+                        else:
+                            raise IllegalItemError(keys, key)
                 else:
-                    raise IllegalItemError(keys, key)
+                    speckey = key
+                specval = spec[speckey]
+                if isinstance(value, dict) and isinstance(specval, dict):
+                    # Item is dict, push to queue
+                    queue.append([value, specval, keys + [key]])
+                elif value is not None and not isinstance(specval, dict):
+                    # Item is value, coerce according to value type
+                    cfg[key] = self.coercers[specval[0]](value, keys + [key])
+                    # [vtype, option_default, option_2, option_3, ...]
+                    if len(specval) > 2:
+                        voptions = specval[1:]
+                        if (isinstance(cfg[key], list) and
+                                any(val not in voptions for val in cfg[key]) or
+                                not isinstance(cfg[key], list) and
+                                cfg[key] not in voptions):
+                            raise IllegalValueError(
+                                'option', keys + [key], cfg[key])
+
+    @classmethod
+    def _coerce_boolean(cls, value, keys):
+        """Coerce value to a boolean."""
+        value = cls._strip_and_unquote(keys, value)
+        if value in ['True', 'true']:
+            return True
+        elif value in ['False', 'false']:
+            return False
+        elif value in ['', None]:
+            return None
         else:
-            speckey = key
-        specval = spec[speckey]
-        if isinstance(val, dict) and isinstance(specval, dict):
-            validate(val, spec[speckey], keys + [key])
-        elif val is not None and not isinstance(specval, dict):
-            # (if val is null we're only checking item validity)
-            cfig[key] = spec[speckey].check(val, keys + [key])
-        else:
-            # raise IllegalItemError(keys, key)
-            # THIS IS OK: blank value
-            # TODO - ANY OTHER POSSIBILITIES?
-            # print 'VAL:', val, '::', keys + [key]
-            pass
-
-
-def check_compulsory(cfig, spec, keys=None):
-    """Check compulsory items are defined in cfig."""
-    if not keys:
-        keys = []
-    for key, val in spec.items():
-        if isinstance(val, dict):
-            check_compulsory(cfig, spec[key], keys + [key])
-        else:
-            if val.args['compulsory']:
-                cfg = cfig
+            raise IllegalValueError('boolean', keys, value)
+
+    @classmethod
+    def _coerce_cycle_point(cls, value, keys):
```

Comment:
Are we able to use the `cylc.cycling.iso8601.SuiteSpecifics.iso8601_parsers[0]` instance of `TimePointParser` rather than creating a new one here or is that not available at this stage?

Reply:
It is not yet available at this stage, and will introduce a circular dependency between parsec and cylc libraries. Having said, I'll take a look to see whether it is possible to do it the other way round or not.

Reply:
(Having looked at it, it may be more trouble than it is worth.)